### ISSUE TITLE
eslint throwing "cannot read property 'name' of null"

### ISSUE THREAD

**Issue by pietrofxq** [eslint/typescript-eslint-parser#593]
**What version of TypeScript are you using?**
3.1.1

**What version of `typescript-eslint-parser` are you using?**
21.0.2

Eslint version: 5.12.0

**What code were you trying to parse?**

```ts
declare module "@ajs/context-path" {
  export default function (): string;
}
```

and 

```js
const BaseState = ({ title, hint, src, ...props }) => (
  <MessageBox {...props}>
    <MessageImage src={src} />
    <MessageText>
      <MainMessage>{title}</MainMessage>
      <p>{hint}</p>
    </MessageText>
  </MessageBox>
)
```

**What did you expect to happen?**
No eslint errors

**What happened?**
![image](https://user-images.githubusercontent.com/6238029/50793077-4d871100-12c7-11e9-9a3d-27ae54a46f80.png)

.eslintrc:

```
{
  "extends": [
    "eslint:recommended",
    "plugin:react/recommended",
    "@atlassian/atlassian-fecq",
    "prettier"
  ],
  "settings": {
    "react": {
      "pragma": "React",
      "version": "^16.5.2"
    }
  },
  "parser": "typescript-eslint-parser",
  "env": {
    "es6": true,
    "browser": true,
    "node": true
  },
  "globals": {
    "I18n": true
  },
  "plugins": [
    "react",
    "typescript"
  ],
  "parserOptions": {
    "sourceType": "module",
    "ecmaFeatures": {
      "experimentalObjectRestSpread": true,
      "jsx": true
    }
  }
}
```

Am I missing something? Using typescript-eslint-parser disables jsx parsing even though I have `jsx:true` in `parserOptions` and throws this `cannot read property 'name' of null` error.

**Comment by armano2** [comment]
```json
"parserOptions": {
    "sourceType": "module",
    "jsx": true
  }
```

**Comment by pietrofxq** [comment]
@armano2 thanks... that got rid of the react issues. Now I still have "cannot read property 'name' of null"

**Comment by pietrofxq** [comment]
This seems to be an issue with `vscode-eslint`. Running `eslint` in the command line doesn't give me such error, it only shows up in the vscode 'problems' panel

**Comment by armano2** [comment]
can you try seeting up 
```json
{
  "eslint.options": { "configFile": "path/.eslintrc.json" }
}
```
looks like its not using `typescript-eslint-parser`

**Comment by pietrofxq** [comment]
same issue :/ I can see that vscode is reading the file because when I change `typescript-eslint-parser` to `babel-eslint` the error changes instantly. Looks like everything is working properly despite of the error showing up in the "problems" tab, I'll live with it for now

**Comment by armano2** [comment]
i was able to reproduce it in console :)

**Comment by armano2** [comment]
ok issue is here https://github.com/eslint/typescript-eslint-parser/blob/master/analyze-scope.js#L323

there is no support for anonymous functions :)

**Comment by JamesHenry** [comment]
Should hopefully be fixed by #596

**Comment by armano2** [comment]
to be precise by this commit: https://github.com/eslint/typescript-eslint-parser/pull/596/commits/08f5a74d85c4314ed9039e2afcce54963187bfca
